NitroxClient/MonoBehaviours/Multiplayer.cs

```diff
@@ -15,6 +16,7 @@
 
 namespace NitroxClient.MonoBehaviours
 {
+    //This class is getting really big and is taking on many responsibilities. It might be worth a joint effort to see if we can plan some refactoring to this guy at some point in the future.
```

Comment:
Yes it would be best to join related `Logic` and `Multiplayer` part together into one class (for AI, Player, Vehicle, etc). Hopefully with a clear distinction between local changes and Nitrox MP addition like RemotePlayer class and the PacketProcessors. I agree that a refactor of this magnitude deserves a design document.

Reply:
I assume it would be solved perfectly with the IoC container @MadMaxOfYore has been mentioning. You have both been added to the private discord channels, where we can do some initial direct discussions.

I never liked the Logic classes though. Most of them don't really do anything other than passing the parameters (usually in the same order) to a packet constructor and sending it using the `packetSender`.

Reply:
Yup, do a Google search for "zenject cheatsheet." The static Container object would become our go-to for all famework dependencies. We can literally drop that in right now and divert everything from Multiplayer.Logic to Container.Resolve<>

It would even handle mapping packets to processors (i think) so that we don't need to keep adding more entries in that static processorParameters collection. The only thing that would prevent that is if zenject does not support generic interface registrations - which almost all DI frameworks do. Oddly enough - the only one of the market that I know for sure does NOT support generic registrations is Microsoft's .NET DI framework - which is ALSO called... Unity -_-

Reply:
I don't think you meant to say "mapping packets to processors" because that's already dealt with statically by the generic argument (every packet needs exactly one packet processor). ProcessorParameters is a completely different mechanism: I wrote it, and it's basically an ugly homemade implementation of what these dependency injection frameworks can do for us (as it simply provides the processor handles to the `Logic` (or related) objects that it needs access to).

I don't think we need generic mappings, since we only need access to non-generic `Logic` classes from within the generic packet handlers.
That is, unless you also want to rewrite the packet handler detection mechanism (and if I'm correct, this sounds like the strategy pattern, where it maps every type of packet to it's own handler).

Reply:
Yup - we'd use the IoC container to change the "strategy" needed to handle the received packet. 

However, we *could* create a Registrar that scans the assembly for processor types like you currently do. Then, using the client as an example, we'd register the concrete processor within the container as a ClientPacketProcessor<Packet>. Then, we'd just resolve ClientPacketProcessor<*ReceivedPacket*> and get a fully initialized instance of the desired Processor.

That would get rid of a lot of code from our code base and create a consistent mechanism for constructing required dependencies.

Reply:
If that's easy to do in the IoC framework without too much extra code (contrary to the current manual implementation that checks the generic parameters etc), I'm all for it :+1: